src/cryptsetup/cryptsetup-token-systemd-tpm2/cryptsetup-token-systemd-tpm2.c

```diff
@@ -0,0 +1,223 @@
+
+        assert(token >= 0);
+
+        /* This must not fail at this moment (internal error) */
+        r = crypt_token_json_get(cd, token, &json);
+        assert(token == r);
+        assert(json);
+
+        if (usrptr)
+                params = *(tpm2_params *)usrptr;
+
+        r = parse_luks2_tpm2_data(json, params.search_pcr_mask, &pcr_mask, &base64_blob, &hex_policy_hash);
+        if (r < 0)
+                return log_debug_open_error(cd, r);
+
+        /* should not happen since cryptsetup_token_validate have passed */
+        r = unbase64mem(base64_blob, SIZE_MAX, &blob, &blob_size);
+        if (r < 0)
+                return log_debug_open_error(cd, r);
+
+        /* should not happen since cryptsetup_token_validate have passed */
+        r = unhexmem(hex_policy_hash, SIZE_MAX, &policy_hash, &policy_hash_size);
+        if (r < 0)
+                return log_debug_open_error(cd, r);
+
+        r = acquire_luks2_key(
+                        pcr_mask,
+                        params.device,
+                        blob,
+                        blob_size,
+                        policy_hash,
+                        policy_hash_size,
+                        &decrypted_key,
+                        &decrypted_key_size);
+        if (r < 0)
+                return log_debug_open_error(cd, r);
+
+        /* Before using this key as passphrase we base64 encode it, for compat with homed */
+        r = base64mem(decrypted_key, decrypted_key_size, &base64_encoded);
+        if (r < 0)
+                return log_debug_open_error(cd, r);
+
+        /* free'd automaticaly by libcryptsetup */
+        *password_len = strlen(base64_encoded);
+        *password = TAKE_PTR(base64_encoded);
+
+        return 0;
+}
+
+/*
+ * libcryptsetup callback for memory deallocation of 'password' parameter passed in
+ * any crypt_token_open_* plugin function
+ */
+_public_ void cryptsetup_token_buffer_free(void *buffer, size_t buffer_len) {
+        erase_and_free(buffer);
+}
+
+/*
+ * prints systemd-tpm2 token content in crypt_dump().
+ * 'type' and 'keyslots' fields are printed by libcryptsetup
+ */
+_public_ void cryptsetup_token_dump(
+                struct crypt_device *cd /* is always LUKS2 context */,
+                const char *json /* validated 'systemd-tpm2' token if cryptsetup_token_validate is defined */) {
+
+        bool first;
+        uint32_t i, pcr, pcr_mask;
+        _cleanup_free_ char* base64_blob = NULL, *hex_policy_hash = NULL;
+
+        if (parse_luks2_tpm2_data(json, UINT32_MAX, &pcr_mask, &base64_blob, &hex_policy_hash))
+                return;
+
+        crypt_log_std(cd, "\ttpm2-pcrs:  ");
+        for (i = 0, first = true; i < TPM2_PCRS_MAX; i++) {
+                pcr = pcr_mask & (UINT32_C(1) << i);
+                if (pcr) {
+                        crypt_log_std(cd, "%s%" PRIu32, first ? "" : ", ", pcr);
+                        first = false;
+                }
+        }
```

Comment:
so for systemd's internal logging we generally have the rule, that strings should be joined before passing them to the log function, so that each log line maps exactly to one log function call.

Thus, maybe use a loop around `strextendf_with_separator()` to concatenate this, it should make this very short and pretty. Then output hte generated string in one go.

(I am mostly thinking heere that some code plus their own logging behind cryptetup's log glue, and would also like to have one log line per log call)